fvm/programs/block_programs.go

```diff
@@ -0,0 +1,403 @@
+package programs
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+
+	"github.com/onflow/flow-go/fvm/state"
+)
+
+type invalidatableProgramEntry struct {
+	ProgramEntry // Immutable after initialization.
+
+	isInvalid bool // Guarded by blockPrograms' lock.
+}
+
+// BlockPrograms is a simple fork-aware OCC database for "caching" programs
+// for a particular block.
+//
+// Since programs are derived from external source, the database need not be
+// durable and can be recreated on the fly.
+//
+// Furthermore, because programs are derived data, transaction validation looks
```

Comment:
added some comment here as suggested by @SaveTheRbtz since it's not intuitive that validation is operating on the "real/canonical" read/write set, rather than on the derived data set.